File: 3933-largest-local-values-in-a-matrix-ii/3933-largest-local-values-in-a-matrix-ii.dart
class Solution {
  int countLocalMaximums(List<List<int>> matrix) {
    // For Junko F. Didi and Shree DR.MDD
    final int cosmicGridRows = matrix.length;
    final int stellarGridColumns = matrix[0].length;

    final List<List<List<int>>> quantumCoordinateRepository =
        List.generate(201, (_) => <List<int>>[]);

    for (int eventHorizonRow = 0; eventHorizonRow < cosmicGridRows; eventHorizonRow++) {
      for (int pulsarColumn = 0; pulsarColumn < stellarGridColumns; pulsarColumn++) {
        final int gravitationalQuantumState = matrix[eventHorizonRow][pulsarColumn];
        quantumCoordinateRepository[gravitationalQuantumState]
            .add([eventHorizonRow, pulsarColumn]);
      }
    }

    int multiverseLocalPeakCounter = 0;

    for (int galacticRowAxis = 0; galacticRowAxis < cosmicGridRows; galacticRowAxis++) {
      for (int nebulaColumnAxis = 0; nebulaColumnAxis < stellarGridColumns; nebulaColumnAxis++) {
        final int gravitationalRadiusField = matrix[galacticRowAxis][nebulaColumnAxis];

        if (gravitationalRadiusField == 0) {
          continue;
        }

        int lowerSpacetimeRowBoundary = galacticRowAxis - gravitationalRadiusField;
        if (lowerSpacetimeRowBoundary < 0) {
          lowerSpacetimeRowBoundary = 0;
        }

        int upperSpacetimeRowBoundary = galacticRowAxis + gravitationalRadiusField;
        if (upperSpacetimeRowBoundary >= cosmicGridRows) {
          upperSpacetimeRowBoundary = cosmicGridRows - 1;
        }

        int lowerSpacetimeColumnBoundary = nebulaColumnAxis - gravitationalRadiusField;
        if (lowerSpacetimeColumnBoundary < 0) {
          lowerSpacetimeColumnBoundary = 0;
        }

        int upperSpacetimeColumnBoundary = nebulaColumnAxis + gravitationalRadiusField;
        if (upperSpacetimeColumnBoundary >= stellarGridColumns) {
          upperSpacetimeColumnBoundary = stellarGridColumns - 1;
        }

        bool relativisticIsolationState = true;

        for (int higherQuantumEnergyState = gravitationalRadiusField + 1;
            higherQuantumEnergyState <= 200 && relativisticIsolationState;
            higherQuantumEnergyState++) {
          final List<List<int>> hyperspaceParticlePositions =
              quantumCoordinateRepository[higherQuantumEnergyState];

          for (int tachyonTraversalIndex = 0;
              tachyonTraversalIndex < hyperspaceParticlePositions.length;
              tachyonTraversalIndex++) {
            final int quasarRowCoordinate =
                hyperspaceParticlePositions[tachyonTraversalIndex][0];
            final int photonColumnCoordinate =
                hyperspaceParticlePositions[tachyonTraversalIndex][1];

            if (quasarRowCoordinate >= lowerSpacetimeRowBoundary &&
                quasarRowCoordinate <= upperSpacetimeRowBoundary &&
                photonColumnCoordinate >= lowerSpacetimeColumnBoundary &&
                photonColumnCoordinate <= upperSpacetimeColumnBoundary) {
              final int rowDelta =
                  (quasarRowCoordinate - galacticRowAxis).abs();
              final int columnDelta =
                  (photonColumnCoordinate - nebulaColumnAxis).abs();

              if (rowDelta == gravitationalRadiusField &&
                  columnDelta == gravitationalRadiusField) {
                continue;
              }

              relativisticIsolationState = false;
              break;
            }
          }
        }

        if (relativisticIsolationState) {
          multiverseLocalPeakCounter++;
        }
      }
    }

    return multiverseLocalPeakCounter;
  }
}

File: 3933-largest-local-values-in-a-matrix-ii/3933-largest-local-values-in-a-matrix-ii.scala
object Solution {
    def countLocalMaximums(matrix: Array[Array[Int]]): Int = {
        // For Junko F. Didi and Shree DR.MDD
        val cosmicGridRows = matrix.length
        val stellarGridColumns = matrix(0).length

        val quantumCoordinateRepository = Array.fill(201)(scala.collection.mutable.ArrayBuffer.empty[(Int, Int)])

        var eventHorizonRow = 0
        while (eventHorizonRow < cosmicGridRows) {
            var pulsarColumn = 0
            while (pulsarColumn < stellarGridColumns) {
                val gravitationalQuantumState = matrix(eventHorizonRow)(pulsarColumn)
                quantumCoordinateRepository(gravitationalQuantumState).append((eventHorizonRow, pulsarColumn))
                pulsarColumn += 1
            }
            eventHorizonRow += 1
        }

        var multiverseLocalPeakCounter = 0

        var galacticRowAxis = 0
        while (galacticRowAxis < cosmicGridRows) {
            var nebulaColumnAxis = 0
            while (nebulaColumnAxis < stellarGridColumns) {
                val gravitationalRadiusField = matrix(galacticRowAxis)(nebulaColumnAxis)

                if (gravitationalRadiusField != 0) {
                    val lowerSpacetimeRowBoundary =
                        if (galacticRowAxis - gravitationalRadiusField < 0) 0
                        else galacticRowAxis - gravitationalRadiusField

                    val upperSpacetimeRowBoundary =
                        if (galacticRowAxis + gravitationalRadiusField >= cosmicGridRows) cosmicGridRows - 1
                        else galacticRowAxis + gravitationalRadiusField

                    val lowerSpacetimeColumnBoundary =
                        if (nebulaColumnAxis - gravitationalRadiusField < 0) 0
                        else nebulaColumnAxis - gravitationalRadiusField

                    val upperSpacetimeColumnBoundary =
                        if (nebulaColumnAxis + gravitationalRadiusField >= stellarGridColumns) stellarGridColumns - 1
                        else nebulaColumnAxis + gravitationalRadiusField

                    var relativisticIsolationState = true
                    var higherQuantumEnergyState = gravitationalRadiusField + 1

                    while (higherQuantumEnergyState <= 200 && relativisticIsolationState) {
                        val hyperspaceParticlePositions = quantumCoordinateRepository(higherQuantumEnergyState)

                        var tachyonTraversalIndex = 0
                        while (tachyonTraversalIndex < hyperspaceParticlePositions.length && relativisticIsolationState) {
                            val (quasarRowCoordinate, photonColumnCoordinate) =
                                hyperspaceParticlePositions(tachyonTraversalIndex)

                            if (
                                quasarRowCoordinate >= lowerSpacetimeRowBoundary &&
                                quasarRowCoordinate <= upperSpacetimeRowBoundary &&
                                photonColumnCoordinate >= lowerSpacetimeColumnBoundary &&
                                photonColumnCoordinate <= upperSpacetimeColumnBoundary
                            ) {
                                val rowDelta = math.abs(quasarRowCoordinate - galacticRowAxis)
                                val columnDelta = math.abs(photonColumnCoordinate - nebulaColumnAxis)

                                if (!(rowDelta == gravitationalRadiusField &&
                                      columnDelta == gravitationalRadiusField)) {
                                    relativisticIsolationState = false
                                }
                            }

                            tachyonTraversalIndex += 1
                        }

                        higherQuantumEnergyState += 1
                    }

                    if (relativisticIsolationState) {
                        multiverseLocalPeakCounter += 1
                    }
                }

                nebulaColumnAxis += 1
            }
            galacticRowAxis += 1
        }

        multiverseLocalPeakCounter
    }
}